docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+diff it against the one they already support, and work their way
+through the changes (see below).
+
+The client will call `GET /api-versions` and pick any number in the
+intersection of the versions it supports and the ones the server
+supports (usually the largest).
+
+*Corner case:* if we want to distinguish between backend-to-backend
+and client-to-backend, we can do that in path suffixes (`GET
+/api-versions/{client,federation}` etc.).
+
+
+### No shared api version
+
+If the intersection set is empty, the client has no way of talking to
+the server.  It needs to politely fail with an error message about
+upgrading, downgrading, or talking to another server.
+
+This should only happen if the distance between last upgrade on client
+and server exceeds the agreed-upon limits (eg., 6 months).
+
+
+### Update detection and version re-negotation
+
+If the server is upgraded and some old supported versions are phased
+out, the client may be caught by surprise.
+
+Servers should respond with status `404`, error label
+`unsupported-version`.  The versioning middleware can do that (see
+above).
+
+Client should install a catch-all that will handle this specific
+error, re-fetch `/api-versions`, and try again.
+
+This will only happen if backend and client grow apart in time beyond
+the supported limit, and there is some chance it will result in an
+empty set of compatible versions, so it's also ok to just fail here.
+
+
+## Strongly typed versions
+
+If we make version an ADT `ApiVersion`, we can remove old versions
+from it in one place and have the compiler guide us through all the
+places where we need to remove it.
+
+There are at least two ways to implement this:
+
+1. Add a few extra servant instances for `(v :: ApiVersion) :> route`.
+2. Define a type family `Versions` that maps `V*` to `"v*"`, and write
+`Versions version :>` in the routing type instead of `"v1"`.
+
+2 seems a lot less work to write, read, and understand.
+
+
+## Data migration (aka data marshalling)
+
+If the shape of an end-point changes between versions (if a data type
+in the routing table becomes a type family), it is often possible to
+write marshalling functions that translate a value from an older
+version into one of a newer version or vice versa.
+
+These functions are called marshalling functions and are useful to
+define separately to keep the application logic clean.
+
+For certain changes to a data type used in an API, marshalling is
+straight-forward in both directions.  The most common example is
+adding an optional attribute to a JSON object:
+
+- *backward migration*: remove the new attribute.
+- *forward migration*: set the new attribute to `null`.
+
+(This is what wire has traditionally done to accomplish client
+backwards compatibility without any API versioning.)
+
+If a mandatory attribute is added in a newer version, there may be a
+plausible default value that can be used in the forward migration
+(backward migration would still remove the field).
```

Comment:
Older clients won't be able to read the field because they are unaware of it.

Reply:
This is not about an additional field that clients can not yet read, but an additional fiend that the backend expects FROM the client. If the client did not specify that field, the backend will assume a default value.